File: pawnote/src/models/portals/student_login_portal.ts
import { Instance } from "../instance";
import { PendingLogin } from "../pending_login";
import { Student } from "../user/student";
import { Webspace } from "../webspace";
import { LoginPortal } from "./login_portal";

export class StudentLoginPortal extends LoginPortal {
	public constructor(instance: Instance) {
		super(instance);
	}

	public async credentials(
		username: string,
		password: string,
		deviceUUID = crypto.randomUUID() as string,
		navigatorIdentifier: string | null = null,
	): Promise<PendingLogin> {
		return super._credentials(Webspace.Students, username, password, deviceUUID, navigatorIdentifier);
	}

	public async token(
		username: string,
		token: string,
		deviceUUID: string,
		navigatorIdentifier: string | null = null,
	): Promise<PendingLogin> {
		return super._token(Webspace.Students, username, token, deviceUUID, navigatorIdentifier);
	}

	public async finish(login: PendingLogin): Promise<Student> {
		return new Student(await super._finish(login), login._session, login._parameters, login._authentication);
	}
}

File: rust/pawnote/src/model/portals/student_login_portal.rs
use std::error::Error;

use uuid::Uuid;

use crate::{Instance, LoginPortal, PendingLogin, Webspace};

pub struct StudentLoginPortal {
    instance: Instance,
}

impl StudentLoginPortal {
    pub fn new(instance: Instance) -> Self {
        Self { instance }
    }

    pub fn instance(&self) -> &Instance {
        &self.instance
    }

    pub async fn credentials(
        &self,
        username: &str,
        password: &str,
        device_uuid: Option<String>,
        navigator_identifier: Option<String>,
    ) -> Result<PendingLogin, Box<dyn Error>> {
        let device_uuid = device_uuid.unwrap_or_else(|| Uuid::new_v4().to_string());

        self._credentials(
            Webspace::Students,
            username,
            password,
            &device_uuid,
            navigator_identifier.as_deref(),
        )
        .await
    }
}

#[async_trait::async_trait]
impl LoginPortal for StudentLoginPortal {
    fn instance(&self) -> &Instance {
        &self.instance
    }
}